include/dlaf/cuda/executor.h

```diff
@@ -0,0 +1,195 @@
+      DLAF_CUDA_CALL(cudaStreamCreateWithPriority(&s, cudaStreamNonBlocking, stream_priority));
+    }
+  }
+
+  StreamPoolImpl& operator=(StreamPoolImpl&&) = default;
+  StreamPoolImpl(StreamPoolImpl&&) = default;
+  StreamPoolImpl(const StreamPoolImpl&) = delete;
+  StreamPoolImpl& operator=(const StreamPoolImpl&) = delete;
+
+  ~StreamPoolImpl() {
+    for (auto& s : streams_) {
+      DLAF_CUDA_CALL(cudaStreamDestroy(s));
+    }
+  }
+
+  cudaStream_t getNextStream() {
+    // Set the device corresponding to the CUBLAS handle.
+    //
+    // The CUBLAS library context is tied to the current CUDA device [1]. A previous task scheduled on
+    // the same thread may have set a different device, this makes sure the correct device is used. The
+    // function is considered very low overhead call [2].
+    //
+    // [1]: https://docs.nvidia.com/cuda/cublas/index.html#cublascreate
+    // [2]: CUDA Runtime API, section 5.1 Device Management
+    DLAF_CUDA_CALL(cudaSetDevice(device_));
+    const std::size_t worker_thread_num = hpx::get_worker_thread_num();
+    DLAF_ASSERT(worker_thread_num != std::size_t(-1), "");
+    std::size_t stream_idx =
+        worker_thread_num * num_streams_per_worker_thread_ +
+        (++current_stream_idxs_[worker_thread_num].data_ % num_streams_per_worker_thread_);
+
+    return streams_[stream_idx];
+  }
+
+  int getDevice() {
+    return device_;
+  }
+};
+}
+
+/// A pool of CUDA streams with reference semantics (copying points to the same
+/// underlying CUDA streams, last reference destroys the references).  Allows
+/// access to CUDA streams in a round-robin fashion.  Each HPX worker thread is
+/// assigned a set of thread local CUDA streams.
+class StreamPool {
+  std::shared_ptr<internal::StreamPoolImpl> streams_ptr_;
+
+public:
+  StreamPool(int device = 0, std::size_t num_streams_per_worker_thread = 3,
+             hpx::threads::thread_priority hpx_thread_priority = hpx::threads::thread_priority::default_)
+      : streams_ptr_(std::make_shared<internal::StreamPoolImpl>(device, num_streams_per_worker_thread,
+                                                                hpx_thread_priority)) {}
+
+  cudaStream_t getNextStream() {
+    DLAF_ASSERT(bool(streams_ptr_), "");
+    return streams_ptr_->getNextStream();
+  }
+
+  int getDevice() {
+    DLAF_ASSERT(bool(streams_ptr_), "");
+    return streams_ptr_->getDevice();
+  }
+
+  bool operator==(StreamPool const& rhs) const noexcept {
+    return streams_ptr_ == rhs.streams_ptr_;
+  }
+
+  bool operator!=(StreamPool const& rhs) const noexcept {
+    return !(*this == rhs);
+  }
+};
+
+/// An executor for cuBLAS functions. Uses handles and streams from the given
+/// HandlePool and StreamPool. A cuBLAS function is defined as any function that
+/// takes a cuBLAS handle as the first argument. The executor inserts a cuBLAS
+/// handle into the argument list, i.e. a handle should not be provided at the
+/// apply/async/dataflow invocation site.
```

Comment:
This comment seems misplaced. This is the CUDA executor.

Reply:
Indeed, good catch!